internal/pkg/cli/recorder/recorder.go

```diff
@@ -0,0 +1,325 @@
+	}
+
+	const programName = "sys_enter"
+	log.Print("Getting bpf program " + programName)
+	program, err := module.GetProgram(programName)
+	if err != nil {
+		return fmt.Errorf("get %s program: %w", programName, err)
+	}
+
+	log.Print("Attaching bpf tracepoint")
+	if _, err := program.AttachTracepoint("raw_syscalls", programName); err != nil {
+		return fmt.Errorf("attach tracepoint: %w", err)
+	}
+
+	log.Print("Getting pids map")
+	r.pids, err = module.GetMap("pids")
+	if err != nil {
+		return fmt.Errorf("get syscalls map: %w", err)
+	}
+
+	log.Print("Module successfully loaded")
+	return nil
+}
+
+func (r *Recorder) unload() {
+	log.Print("Unloading bpf module")
+	r.pids.GetModule().Close()
+	os.RemoveAll(r.btfPath)
+}
+
+func (r *Recorder) findBtfPath() (string, error) {
+	// Use the system btf if possible
+	if _, err := os.Stat("/sys/kernel/btf/vmlinux"); err == nil {
+		log.Print("Using system btf file")
+		return "", nil
+	}
+
+	log.Print("Trying to find matching in-memory btf")
+
+	btf := types.Btf{}
+	if err := json.Unmarshal([]byte(btfJSON), &btf); err != nil {
+		return "", fmt.Errorf("unmarshal btf JSON: %w", err)
+	}
+
+	res, err := osrelease.Read()
+	if err != nil {
+		return "", fmt.Errorf("read os-release file: %w", err)
+	}
+
+	osID := types.Os(res["ID"])
+	btfOs, ok := btf[osID]
+	if !ok {
+		log.Printf("OS not found in btf map: %s", osID)
+		return "", nil
+	}
+	log.Printf("OS found in btf map: %s", osID)
+
+	osVersion := types.OsVersion(res["VERSION_ID"])
+	btfOsVersion, ok := btfOs[osVersion]
+	if !ok {
+		log.Printf("OS version not found in btf map: %s", osVersion)
+		return "", nil
+	}
+	log.Printf("OS version found in btf map: %s", osVersion)
+
+	uname := syscall.Utsname{}
+	if err := syscall.Uname(&uname); err != nil {
+		return "", fmt.Errorf("uname syscall failed: %w", err)
+	}
+
+	arch := types.Arch(util.ToStringInt8(uname.Machine))
+	btfArch, ok := btfOsVersion[arch]
+	if !ok {
+		log.Printf("Architecture not found in btf map: %s", arch)
+		return "", nil
+	}
+	log.Printf("Architecture found in btf map: %s", arch)
+
+	release := util.ToStringInt8(uname.Release)
+	kernel := types.Kernel(release)
+	btfBytes, ok := btfArch[kernel]
+	if !ok {
+		log.Printf("Kernel not found in btf map: %s", kernel)
+		return "", nil
+	}
+	log.Printf("Kernel found in btf map: %s", kernel)
+
+	file, err := os.CreateTemp(
+		"",
+		fmt.Sprintf("spo-btf-%s-%s-%s-%s", osID, osVersion, arch, kernel),
+	)
+	if err != nil {
+		return "", fmt.Errorf("create temp file: %w", err)
+	}
+	defer file.Close()
+
+	if _, err := file.Write(btfBytes); err != nil {
+		return "", fmt.Errorf("write BTF: %w", err)
+	}
+
+	log.Printf("Wrote BTF to file: %s", file.Name())
+	return file.Name(), nil
+}
+
+func (r *Recorder) processData(pid uint32, cmdName, filename string) error {
+	log.Printf("Processing recorded data")
+
+	iterator := r.pids.Iterator()
+	for iterator.Next() {
+		currentPid := binary.LittleEndian.Uint32(iterator.Key())
+
+		if currentPid != pid {
```

Comment:
I am wondering, if it is not more memory efficient to use the program name instead. This approach will record the syscalls of all pids running in the system during the profile recording. 

I am thinking something along these lines:
- create another map to hold the program name as input to eBPF program
- write the program name from Go/userspace into that map
- start the program execution
- record only syscalls belonging to the program name by filter anything out

There is a risk when multiple instances of the same program name runs in parallel, but overall there will be a lot less syscalls recorded.

I am not sure if this optimisation is worth implementing since is just a tool.

Reply:
Hm, we could use the program name as rodata constant for additional filtering. I'm going to follow-up on that as well.